File: app/src/main/java/com/pkfcooperparry/template/ui/form/page/PageAdapter.java
package com.pkfcooperparry.template.ui.form.page;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.pkfcooperparry.template.R;
import com.pkfcooperparry.template.data.model.db.visit.Page;

import java.util.List;

public class PageAdapter extends RecyclerView.Adapter<PageAdapter.ViewHolder> {

    Context context;
    private List<Page> pages;
    private LayoutInflater inflater;
    private ItemClickListener itemClickListener;

    // data is passed into the constructor
    public PageAdapter(Context context, ItemClickListener itemClickListener, List<Page> pages) {
        this.inflater = LayoutInflater.from(context);
        this.itemClickListener = itemClickListener;
        this.context = context;
        this.pages = pages;
    }

    // inflates the row layout from xml when needed
    @Override
    @NonNull
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = inflater.inflate(R.layout.row_page, parent, false);
        return new ViewHolder(view);
    }

    // binds the data to the view and textview in each row
    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        holder.name.setTextColor(ContextCompat.getColor(context, R.color.page_text_color));

        Page page = pages.get(position);
        holder.name.setText(page.getPageTitle());
        holder.image.setImageDrawable(getImage(page));
        holder.parent.setOnClickListener(v -> {
            if (itemClickListener != null) {
                itemClickListener.onItemClick(page);
            }
        });
        holder.item.setBackground(ContextCompat.getDrawable(context, R.drawable.bg_page_selector));


    }

    private Drawable getImage(Page page) {
        return ContextCompat.getDrawable(context, R.drawable.page_placeholder);

    }

    // total number of rows
    @Override
    public int getItemCount() {
        return pages.size();
    }

    public void setClickListener(ItemClickListener itemClickListener) {
        this.itemClickListener = itemClickListener;
    }


    public interface ItemClickListener {
        void onItemClick(Page page);

    }

    // stores and recycles views as they are scrolled off screen
    public class ViewHolder extends RecyclerView.ViewHolder {
        TextView name;
        ImageView image;
        RelativeLayout parent;
        RelativeLayout item;

        ViewHolder(View itemView) {
            super(itemView);
            name = itemView.findViewById(R.id.pageName);
            image = itemView.findViewById(R.id.pageIcon);
            parent = itemView.findViewById(R.id.topItem);
            item = itemView.findViewById(R.id.item);

        }

    }
}

File: app/src/main/java/com/pkfcooperparry/template/ui/form/page/PageAdapter.kt
package com.pkfcooperparry.template.ui.form.page

import android.content.Context
import android.graphics.drawable.Drawable
import android.support.v4.content.ContextCompat
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.RelativeLayout
import android.widget.TextView

import com.pkfcooperparry.template.R
import com.pkfcooperparry.template.data.model.db.visit.Page

class PageAdapter// data is passed into the constructor
(internal var context: Context, private var itemClickListener: ItemClickListener?, private val pages: List<Page>) : RecyclerView.Adapter<PageAdapter.ViewHolder>() {
    private val inflater: LayoutInflater

    init {
        this.inflater = LayoutInflater.from(context)
    }

    // inflates the row layout from xml when needed
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = inflater.inflate(R.layout.row_page, parent, false)
        return ViewHolder(view)
    }

    // binds the data to the view and textview in each row
    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        holder.name.setTextColor(ContextCompat.getColor(context, R.color.page_text_color))

        val page = pages[position]
        holder.name.text = page.pageTitle
        holder.image.setImageDrawable(getImage(page))
        holder.parent.setOnClickListener { v ->
            if (itemClickListener != null) {
                itemClickListener!!.onItemClick(page)
            }
        }
        holder.item.background = ContextCompat.getDrawable(context, R.drawable.bg_page_selector)


    }

    private fun getImage(page: Page): Drawable? {
        return ContextCompat.getDrawable(context, R.drawable.page_placeholder)

    }

    // total number of rows
    override fun getItemCount(): Int {
        return pages.size
    }

    fun setClickListener(itemClickListener: ItemClickListener) {
        this.itemClickListener = itemClickListener
    }


    interface ItemClickListener {
        fun onItemClick(page: Page)

    }

    // stores and recycles views as they are scrolled off screen
    inner class ViewHolder internal constructor(itemView: View) : RecyclerView.ViewHolder(itemView) {
        internal var name: TextView
        internal var image: ImageView
        internal var parent: RelativeLayout
        internal var item: RelativeLayout

        init {
            name = itemView.findViewById(R.id.pageName)
            image = itemView.findViewById(R.id.pageIcon)
            parent = itemView.findViewById(R.id.topItem)
            item = itemView.findViewById(R.id.item)

        }

    }
}